src/core/storage/storage.js

```diff
@@ -32,65 +88,162 @@ export default class GlobalStorage {
    * @param {string} key The storage key to set
    * @param {*} data The data to set
    */
-  set (key, data) {}
+  set (key, data) {
+    if (key === undefined || key === null || typeof key !== 'string') {
+      throw new AnswersStorageError('Invalid storage key provided', key, data);
+    }
+
+    if (typeof data === 'undefined') {
+      throw new AnswersStorageError('Invalid storage key provided', key, data);
+    }
+
+    if (this.moduleDataContainer.get(key) === undefined) {
+      this.moduleDataContainer.set(key, new ModuleData(key));
+      this._applyFutureListeners(key);
+    }
+    this.moduleDataContainer.get(key).set(data);
+  }
 
   /**
    * Updates the storage and adds the entry to a persistent storage
-   * buffer. The entry is not yet added to the persistent storage.
+   * buffer. The entry is not added to the persistent storage until
+   * the buffer is flushed.
    *
    * @param {string} key The storage key to set
    * @param {*} data The data to set
    */
-  setWithPersist (key, data, replaceHistory) {}
+  setWithPersist (key, data) {
+    this.set(key, data);
+
+    let serializedData = data;
+    if (typeof data !== 'string') {
+      serializedData = JSON.stringify(data);
+    }
+
+    this.persistentStorageBuffer.set(key, serializedData);
+  }
 
   /**
    * Flushes the persistent storage buffer and adds all entries
    * to the persistent storage.
    */
-  flushPersist () {}
+  flushPersist () {
+    this.persistentStorage = this.persistentStorageBuffer;
+
+    const persistentStorageClone = new DefaultPersistentStorage(this.popListener);
+    this.persistentStorage.getAll().forEach((value, key) => {
+      persistentStorageClone.set(key, value);
+    });
+
+    this.persistentStorageBuffer = persistentStorageClone;
+
+    this.stateUpdateListener(
+      this.persistentStorage.getAll(),
+      this.persistentStorage.getUrlWithCurrentState()
+    );
+  }
 
   /**
    * Get the current state for the provided key
    *
    * @param {string} key The storage key to get
-   * @return {*} The state for the provided key
+   * @return {*} The state for the provided key, undefined if key doesn't exist
    */
-  get (key) {}
+  get (key) {
+    if (!this.moduleDataContainer.get(key)) {
+      return this.moduleDataContainer.get(key);
```

Comment:
would this.moduleDataContainer.has(key) be better (also for other similar usages like in set, on, and off)? could we also just return undefined instead of doing another get?

Reply:
Moved away from module data! Returning the value directly now.